File: c/54_Spiral_Matrix.js
var spiralOrder = function(matrix) {
  if (matrix.length == 0) return [];

  let res = [];
  let top = 0;
  let bottom = matrix.length - 1;
  let left = 0;
  let right = matrix[0].length - 1;

  while (left <= right && top <= bottom) {
    for (let j = left; j <= right; j++) {
      res.push(matrix[top][j]);
    }
    top++;

    for (let i = top; i <= bottom; i++) {
      res.push(matrix[i][right]);
    }
    right--;

    if (left > right || top > bottom) break;

    for (let j = right; j >= left; j--) {
      res.push(matrix[bottom][j]);
    }
    bottom--;

    for (let i = bottom; i >= top; i--) {
      res.push(matrix[i][left]);
    }
    left++;
  }
  return res;
};

File: c/54_Spiral_Matrix.c
#include <stdio.h>
#include <stdlib.h>

int* spiralOrder(int** matrix, int matrixSize, int* matrixColSize,
                 int* returnSize) {
  int* res;
  int m = matrixSize;
  int n = *matrixColSize;
  int idx = 0;
  int top = 0;
  int bottom = m - 1;
  int left = 0;
  int right = n - 1;
  res = (int*)malloc(sizeof(int) * m * n);

  while (left <= right && top <= bottom) {
    for (int j = left; j <= right; j++) {
      res[idx++] = matrix[top][j];
    }
    top++;

    for (int i = top; i <= bottom; i++) {
      res[idx++] = matrix[i][right];
    }
    right--;

    if (left > right || top > bottom) break;

    for (int j = right; j >= left; j--) {
      res[idx++] = matrix[bottom][j];
    }
    bottom--;

    for (int i = bottom; i >= top; i--) {
      res[idx++] = matrix[i][left];
    }
    left++;
  }
  *returnSize = idx;
  return res;
}